src/hypertable_restrict_info.c

```diff
@@ -0,0 +1,351 @@
+	int			i;
+
+	res->capacity = rel->max_attr;
+	for (i = 0; i < ht->space->num_dimensions; i++)
+	{
+		DimensionRestrictInfo *dri = dimension_restrict_info_create(&ht->space->dimensions[i]);
+
+		res->diminson_restriction[AttrNumberGetAttrOffset(ht->space->dimensions[i].column_attno)] = dri;
+	}
+
+	return res;
+}
+
+static DimensionRestrictInfo *
+hypertable_restrict_info_get(HypertableRestrictInfo *hri, int attno)
+{
+	Assert(attno > 0 && AttrNumberGetAttrOffset(attno) < hri->capacity);
+	return hri->diminson_restriction[AttrNumberGetAttrOffset(attno)];
+}
+
+static bool
+hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *root, OpExpr *clause)
+{
+	Expr	   *leftop,
+			   *rightop,
+			   *expr;
+	DimensionRestrictInfo *dri;
+	Var		   *v;
+	Const	   *c;
+	Oid			op_oid;
+	RangeTblEntry *rte;
+	Oid			columntype;
+	TypeCacheEntry *tce;
+	int			strategy;
+	Oid			lefttype,
+				righttype;
+
+	if (list_length(clause->args) != 2)
+		return false;
+
+	/* Same as constraint_exclusion */
+	if (contain_mutable_functions((Node *) clause))
+		return false;
+
+	leftop = (Expr *) get_leftop((Expr *) clause);
+	if (IsA(leftop, RelabelType))
+		leftop = ((RelabelType *) leftop)->arg;
+	rightop = (Expr *) get_rightop((Expr *) clause);
+	if (IsA(rightop, RelabelType))
+		rightop = ((RelabelType *) rightop)->arg;
+
+	if (IsA(leftop, Var))
+	{
+		v = (Var *) leftop;
+		expr = rightop;
+		op_oid = clause->opno;
+	}
+	else if (IsA(rightop, Var))
+	{
+		v = (Var *) rightop;
+		expr = leftop;
+		op_oid = get_commutator(clause->opno);
+	}
+	else
+		return false;
+
+	dri = hypertable_restrict_info_get(hri, v->varattno);
+	/* the attribute is not a dimension */
+	if (dri == NULL)
+		return false;
+
+	expr = (Expr *) eval_const_expressions(root, (Node *) expr);
+
+	if (!IsA(expr, Const) ||!OidIsValid(op_oid) || !op_strict(op_oid))
+		return false;
+
+	c = (Const *) expr;
+
+	rte = rt_fetch(v->varno, root->parse->rtable);
+
+	columntype = get_atttype(rte->relid, dri->dimension->column_attno);
+	tce = lookup_type_cache(columntype, TYPECACHE_BTREE_OPFAMILY);
+
+	if (!op_in_opfamily(op_oid, tce->btree_opf))
+		return false;
+
+	get_op_opfamily_properties(op_oid,
+							   tce->btree_opf,
+							   false,
+							   &strategy,
+							   &lefttype,
+							   &righttype);
+
+	return dimension_restrict_info_add(dri, strategy, c);
+}
+
+static void
+hypertable_restrict_info_add_restrict_info(HypertableRestrictInfo *hri, PlannerInfo *root, RestrictInfo *ri)
+{
+	Expr	   *e = ri->clause;
+
+	if (IsA(e, OpExpr))
+	{
+		if (hypertable_restrict_info_add_op_expr(hri, root, (OpExpr *) e))
+			hri->num_added_restrictions++;
+	}
+}
+
+void
+hypertable_restrict_info_add(HypertableRestrictInfo *hri,
+							 PlannerInfo *root,
+							 List *base_restrict_infos)
+{
+	ListCell   *lc;
+
+	foreach(lc, base_restrict_infos)
+	{
+		RestrictInfo *ri = lfirst(lc);
+
+		hypertable_restrict_info_add_restrict_info(hri, root, ri);
+	}
+}
+
+bool
```

Comment:
I don't think I can make this inline or macro since I don't expose HypertableRestrictInfo internals outside this .c file. Why do you think this needs to be optimized?